Extend connectivity troubleshooting guide (#2108)
kyma-project/istio: docs/user/troubleshooting/03-00-network-connectivity.md

## 1. Global Analysis

As a starting point, run `istioctl analyze` to check the state of your mesh and point out common errors and misconfigurations.

- Analyze a specific namespace:
  ```bash
  istioctl analyze -n <your-namespace>
  ```

- Analyze the entire cluster:
  ```bash
  istioctl analyze --all-namespaces
  ```

What to look for:
- **Errors and Warnings:** Review all output carefully, especially warnings like `IST0133`, which indicate an AuthorizationPolicy might be blocking more traffic than intended. The analysis often includes the name and namespace of the problematic resource.

## 2. AuthorizationPolicies

Misconfigured AuthorizationPolicy resources are a common cause of access issues, such as `403` errors or unexpected connection timeouts.

- List all AuthorizationPolicies in a namespace:
  ```bash
  kubectl get authorizationpolicies -n <your-namespace>
  ```
- List all AuthorizationPolicies cluster-wide:
  ```bash
  kubectl get authorizationpolicies --all-namespaces
  ```
- Describe a specific AuthorizationPolicy to see its details:
  ```bash
  kubectl describe authorizationpolicy <policy-name> -n <namespace>
  ```

What to look for:

> - **Deny policies:** Deny policies are powerful and may block more traffic than intended. A deny rule without the **ports** field on an HTTP-based **hosts** rule can block all TCP traffic, not just HTTP.

Reviewer: To be honest, allow policies are even more powerful :) 
Having only one already enforces that the policy is matched, that is not the case with deny

- **Scope:** Check the **selector** to see which workloads the policy applies to. An empty selector applies the policy to all workloads in the namespace.
- **Rules:** Look at the **to**, **from**, and **when** clauses to understand what traffic is being allowed or denied.

## 3. RequestAuthentications

RequestAuthentication defines how JSON Web Tokens (JWTs) are validated. Misconfigurations often lead to `401 Unauthorized` responses.

- List all request authentications in a namespace:
  ```bash
  kubectl get requestauthentications -n <your-namespace>
  ```
- Describe a specific RequestAuthentication to see its configuration:
  ```bash
  kubectl describe requestauthentication <ra-name> -n <namespace>
  ```

What to look for:

> - **jwksUri or jwks:** Ensure the JSON Web Key Set (JWKS) endpoint is correct and reachable from the Istio ingress gateway.

Reviewer: ```suggestion
- **jwksUri** Ensure the JSON Web Key Set (JWKS) endpoint is correct and reachable from Istiod.
```

- **issuer:** The `iss` claim in the JWT must match this value.

> - **audiences:** The `aud` claim in the JWT must match one of the values here.

Reviewer: ```suggestion
- **audiences:** In case those are set, the `aud` claim in the JWT must match one of the values here.
```

- List all EnvoyFilters cluster-wide:
  ```bash
  kubectl get envoyfilters --all-namespaces
  ```
- Examine a specific filter:
  ```bash
  kubectl get envoyfilter <filter-name> -n <namespace> -o yaml
  ```

What to look for:
- **Workload selector:** Check which Pods the filter applies to.
- **Filters:** Look at the filter configuration. These are often complex Lua scripts or WASM filters. A bug in the script can break traffic.

## 5. Envoy Configuration

If the previous steps don't reveal the issue, inspect the live Envoy configuration for a specific Pod. This shows exactly which rules Envoy applies.

- Get the proxy config for a specific Pod:
  ```bash
  istioctl proxy-config <all|listeners|clusters|routes|endpoints|bootstrap> <pod-name> -n <namespace>
  ```
- For example, check the routes for a Pod:
  ```bash
  istioctl proxy-config routes <pod-name>.<namespace> -o json
  ```

What to look for:
- **Listeners:** Check if your Pod is listening on the expected port.
- **Routes:** Verify that the host and paths are routed to the correct cluster (service). Check for route-level configurations like timeouts or retries.

> - **Clusters:** Ensure the upstream service endpoints are correctly identified.

Reviewer: This section is not helping, as:
- it is hard to identify the listed points anyway
- those will likely be OK
- problems are more often e.g. in authz/authn config
```suggestion
```

## 6. Sidecar Logs

The Istio sidecar (`istio-proxy`) logs can provide direct clues about why a request is failing.

- Get logs from the `istio-proxy` container:
  ```bash
  kubectl logs <pod-name> -n <namespace> -c istio-proxy
  ```
- Filter logs by response code or other keywords:
  ```bash
  kubectl logs <pod-name> -n <namespace> -c istio-proxy | grep "RBAC"
  ```

What to look for:
- **Access denied messages:** Search for `RBAC: access denied`. This is a clear sign that an AuthorizationPolicy is blocking the request. The log entry often includes details about the matched policy.
- **Upstream connection failures:** Look for messages indicating connection errors to upstream services.

> - **JWT errors:** Errors related to JWT validation suggest issues with RequestAuthentication.

Reviewer: Sidecar logs won't be emitted in default configuration
```suggestion
```

  ```bash
  istioctl analyze -n <your-namespace>
  ```

- Analyze the entire cluster:
  ```bash
  istioctl analyze --all-namespaces
  ```

What to look for:
- **Errors and Warnings:** Review all output carefully, especially warnings like `IST0133`, which indicate an AuthorizationPolicy might be blocking more traffic than intended. The analysis often includes the name and namespace of the problematic resource.

## 2. AuthorizationPolicies

Misconfigured AuthorizationPolicy resources are a common cause of access issues, such as `403` errors or unexpected connection timeouts.

- List all AuthorizationPolicies in a namespace:
  ```bash
  kubectl get authorizationpolicies -n <your-namespace>
  ```
- List all AuthorizationPolicies cluster-wide:
  ```bash
  kubectl get authorizationpolicies --all-namespaces
  ```
- Describe a specific AuthorizationPolicy to see its details:
  ```bash
  kubectl describe authorizationpolicy <policy-name> -n <namespace>
  ```

What to look for:
- **Deny policies:** Deny policies may block more traffic than intended. A deny rule without the **ports** field on an HTTP-based **hosts** rule can block all TCP traffic, not just HTTP.

> - **Allow policies:** If a single allow policy exists in a namespace, all traffic that does not match it is denied. Make sure that all intended traffic is covered by allow rules.

Reviewer: ```suggestion
- **Allow policies:** If even a single allow policy matches a workload, all traffic that does not match any allow is denied. Make sure that all intended traffic is covered by allow rules.
```

Reviewer 2: how about "... that does not match any allow rule is denied."?

> - **Allow policies:** Allow policies can also be dangerous. As soon as a single allow policy exists in a namespace, all traffic that does not match it is denied. Make sure that all intended traffic is covered by allow rules.

Reviewer 2: ```suggestion
- **Allow policies:** If a single allow policy exists in a namespace, all traffic that does not match it is denied. Make sure that all intended traffic is covered by allow rules.
```
